paasta_tools/frameworks/native_scheduler.py

```diff
@@ -240,7 +278,15 @@ def tasks_for_offer(self, driver, offer):
             remainingMem -= task_mem
             remainingPorts -= set([task_port])
 
-        return tasks
+        # raise constraint error but only if no other tasks fit/fail the offer
+        if total > 0 and failed_constraints == total:
+            raise ConstraintFailAllTasksError
+
+        if len(tasks) > 0:
+            update_constraint_state(offer, self.constraints,
```

Comment:
Won't we need to update our constraint state (or a temporary copy) every time we add a task to `tasks`? As written, I think this method could end up creating multiple tasks on the same offer and violating a UNIQUE constraint.